examples/bevy_state.rs

```diff
@@ -0,0 +1,120 @@
+use bevy::{
+    prelude::{App as BevyApp, AssetServer, Commands, Res, ResMut, State, SystemSet},
+    window::WindowDescriptor,
+    DefaultPlugins,
+};
+use kayak_ui::bevy::{BevyContext, BevyKayakUIPlugin, FontMapping, UICameraBundle};
+use kayak_ui::core::{
+    render, rsx, widget, Event, EventType, Index, KayakContextRef, KeyCode, OnEvent, WidgetProps,
+};
+use kayak_ui::widgets::{App, Text};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+enum GameState {
+    MainMenu,
+    Options,
+    Play,
+}
+
+fn swap(mut state: ResMut<State<GameState>>) {
+    if *state.current() == GameState::MainMenu {
+        let _ = state.set(GameState::Options);
+    } else if *state.current() == GameState::Options {
+        let _ = state.set(GameState::Play);
+    } else {
+        let _ = state.set(GameState::MainMenu);
+    }
+}
+
+fn handle_input(context: &mut KayakContextRef, event: &mut Event) {
+    match event.event_type {
+        EventType::KeyDown(event) => {
+            if event.key() == KeyCode::Space {
+                context.query_world::<ResMut<State<GameState>>, _, _>(swap);
+            }
+        }
+        _ => {}
+    };
+}
+#[derive(WidgetProps, Default, Debug, PartialEq, Clone)]
+pub struct StateSwitcherProps;
+
+#[widget]
+fn StateSwitcher(props: StateSwitcherProps) {
+    rsx! {
+        <Text content={"Press space to switch states!".to_string()} size={32.0} />
+    }
+}
+
+fn create_main_menu(mut commands: Commands) {
+    let context = BevyContext::new(|context| {
+        render! {
+            <App on_event={Some(OnEvent::new(handle_input))}>
+                <Text content={"Main Menu".to_string()} size={32.0} />
+                <StateSwitcher />
+            </App>
+        }
+    });
+
+    commands.insert_resource(context);
+}
+
+fn create_options_menu(mut commands: Commands) {
+    let context = BevyContext::new(|context| {
+        render! {
+            <App on_event={Some(OnEvent::new(handle_input))}>
+                <Text content={"Options".to_string()} size={32.0} />
+                <StateSwitcher />
+            </App>
+        }
+    });
+
+    commands.insert_resource(context);
+}
+
+fn create_play_menu(mut commands: Commands) {
```

Comment:
The more I think about this the more I dislike it. I think users should instead use scenes here. We should think about how we want to handle that.

Reply:
I think it's okay to do for now. We could maybe explore scenes in another PR/RFC?